### PR TITLE
Avoid unnecessary reloads of store data

### PR DESCRIPTION
**Pull request by alexlarsson**
We're currently reloading the layers json file every time we (ourselves) write it

This PR fixes this.

### CODE REVIEW

**Comment by alexlarsson** [comment]
Hmm, the tests seem to be built on the idea that the first Modified *should* return true. I think this is sort of weird, but I guess we can work with that by calling Modified() and extra time in the store.

**Review by rhatdan — APPROVED** [review]
LGTM
@mtrmac PTAL

File: pkg/ioutils/fswriters.go
Comment (mtrmac) [inline pkg/ioutils/fswriters.go:123]: AFAIK it’s not documented that `w.f.Close()` won’t update the timestamp (e.g. it could be buffering writes for some reason).

OTOH, sure, `Stat()` after closing the file may lose updates. But then all of this protects against ~rogue external writers who don’t lock correctly…

**Review by mtrmac — COMMENTED** [review]
No, `Modified` needs to be dropped entirely, not made more prevalent: https://github.com/containers/storage/issues/1420 .

I think #1436 + #1346 is a superset of this.

**Comment by rhatdan** [comment]
@alexlarsson Needs a rebase/perhaps a rewrite.

File: pkg/ioutils/fswriters.go
Comment (alexlarsson) [reply pkg/ioutils/fswriters.go:123]: I dunno how *documented* this is per se, but that is now how delayed writes work. The mtime information is part of the struct inode, and it will be modified when you do actual vfs operations, not when dirty buffers are flushed to disk.

**Comment by alexlarsson** [comment]
This is not needed anymore on master.

File: pkg/ioutils/fswriters.go
Comment (mtrmac) [reply pkg/ioutils/fswriters.go:123]: I was thinking of the Go `os.File` possibly buffering data, the way `<stdio.h>` traditionally does. I agree it would be surprising if it happened, but nothing explicitly promises that it won’t.

>  The mtime information is part of the struct inode, and it will be modified when you do actual vfs operations, not when dirty buffers are flushed to disk.

(BTW, https://pubs.opengroup.org/onlinepubs/9699919799/ “file times update” says that it might be delayed after the actual write even inside the kernel — in a way that wouldn’t matter here.)

**Comment by alexlarsson** [comment]
Ah, sorry, we still need the second half of this

File: pkg/ioutils/fswriters.go
Comment (alexlarsson) [reply pkg/ioutils/fswriters.go:123]: I mean, it better have pushed all os.File data to the kernel, or the fdatasync() call we just did was completely wasted, having persisted to disk a partial json file...

Anyway, I'm pretty sure that in practice (i.e. on linux) this will work. Even if Xenix 0.0.1 might do something weird and still be according to spec. And anwyay, as you say, if we get the wrong mtime all that happens is that we unnecessarily load the file again.

**Comment by rhatdan** [comment]
LGTM
@mtrmac PTAL

File: pkg/ioutils/fswriters.go
Comment (mtrmac) [reply pkg/ioutils/fswriters.go:123]: > I mean, it better have pushed all os.File data to the kernel, or the fdatasync() call we just did was completely wasted,

Oh, that’s a strong argument.

File: pkg/ioutils/fswriters.go
Comment (mtrmac) [inline pkg/ioutils/fswriters.go:122]: Why wouldn’t this be a fatal error?

File: pkg/ioutils/fswriters.go
Comment (mtrmac) [inline pkg/ioutils/fswriters.go:17]: Returning a value in “options” seems very surprising to me. At the very least, it should be clearly documented.

Going by the “`…Options` is input to the operation” logic, a `WriteModTimeOnSuccess *time.Time` field would make at bit more sense … and, OTOH, be more cumbersome to use.

Ultimately it would be cleanest to add a new function that returns that value (or a struct containing that value, for future expansion?)

File: layers.go
Comment (mtrmac) [inline layers.go:819]: This special case is not documented … and if it represents a `fstat` failure, I don’t think it’s really necessary to handle that this way.

**Review by mtrmac — COMMENTED** [review]
Logic WRT locking LGTM (and with the “pending writes” debate decisively settled, this is a clearly better approach than the one in #1346).

As for the API design, I’ll fully defer to actual c/storage maintainers.

File: pkg/ioutils/fswriters.go
Comment (alexlarsson) [reply pkg/ioutils/fswriters.go:122]: That is a tricky question, 99.9% of all callers don't need that Mod time so might not care about this failure. On the other hand, if this actually fails then something pretty weird is going an and maybe you still want to know.

File: pkg/ioutils/fswriters.go
Comment (alexlarsson) [reply pkg/ioutils/fswriters.go:17]: I agree that this is somewhat weird, but the primary driver of this particular API was to avoid changing the API and also not needlessly complicating the full API. It is weird, I agree, even ugly. But it is pretty effective, and unlikely to bother anyone. The alternative would be to create yet another "even more options" alternative of the function and have everything chain to that. That is clearly aesthetically better, but increases the amount of api, code and indirect calls at runtime.

File: pkg/ioutils/fswriters.go
Comment (alexlarsson) [reply pkg/ioutils/fswriters.go:125]: Is fatal now

**Comment by alexlarsson** [comment]
Minor change, made the fstat() failure fatal and added some docs.

**Review by vrothberg — COMMENTED** [review]
LGTM

File: pkg/ioutils/fswriters.go
Comment (mtrmac) [reply pkg/ioutils/fswriters.go:125]: AFAICS this error is still ignored.

Comment (mtrmac) [reply pkg/ioutils/fswriters.go:125]: I’m sorry, never mind, my mistake. It’s even very clearly documented.

**Pull request merged**
